### PR TITLE
Global views

### PR DESCRIPTION
**Pull request by dimitri-justeau**
Fixes # .

Changes proposed in this PR:
- Extend the view API to allow global views, i.e. views over an arbitrary number of variables. Global views are especially useful to channel different types of variables, such as a set variable observing an array of boolean variables for instance.
- Implement IntsSetView, which is a view-based alternative to the setsIntsChanneling constraint, and BoolsSetView, which is a view-based alternative to the setsBoolsChanneling constraint.
- Introduce SetDynamicFilter, which is a set data structure that does not store data but instead dynamically returns properties and iterators from another object or collection of objects. The concept is more or less the same as views but for data structure. I did not use the name "view" to avoid confusion. Such sets are particularly useful for set views and graph view that need to perform data transformations, as they avoid to recompute an entire set or graph data structure at each bound read operation. This new type of view can also be used to create dynamically filtered sets, such as retrieving only the odds values of a set data structure.
- Comprehensive performances tests should be done to compare these new views to existing channelling constraints, but in my first basic tests I observed a speed up of x2 with view-based channelling compared to propagator-based channelling.
 
@chocoteam/core-developer

### CODE REVIEW

File: solver/src/main/java/org/chocosolver/solver/variables/view/IntsSetView.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+/**
+ * Set view over an array of integer variables defined such that:
+ * with v and offset two integers (constant) intVariables[x - offset] = c <=> x in set.
+ */
+public class IntsSetView<I extends IntVar> extends SetView<I> {
+
+    /**
+     * Integer value such that intVariables[x - offset] = v <=> x in set
+     */
+    private int v;
```
[the remark this answers is not in the document]
Comment (dimitri-justeau) [reply solver/src/main/java/org/chocosolver/solver/variables/view/IntsSetView.java:33]: Totally, this is a good way to generalize this view a bit more. I'm adding this in a new commit.

File: solver/src/main/java/org/chocosolver/solver/variables/view/IView.java [hunk 2]
```diff
@@ -27,17 +27,32 @@
...
 public interface IView<V extends Variable> extends ICause, Variable {
 
     /**
-     * Return the basis variable
+     * Return the array of observed variables
      *
-     * @return variable observed
+     * @return variables observed
+     */
+    V[] getVariables();
```
[the remark this answers is not in the document]
Comment (dimitri-justeau) [reply solver/src/main/java/org/chocosolver/solver/variables/view/IView.java:48]: I moved the `V getVariable()` method in the `IntView` abstract class, that is extended by every previously existing view. In classes such as `Task` and `ClauseBuilded`, specific classes of view were used, derived from `IntView`. The `getVariable()` method is thus still accessible from these classes. In accordance with the previous comment, defining an interface for unary view could remove any ambiguity, what do you think of this ?

File: solver/src/main/java/org/chocosolver/solver/variables/view/IntView.java [hunk 3]
```diff
@@ -401,7 +398,7 @@ public void createDelta() {
     }
 
     @Override
-    public void notify(IEventType event) throws ContradictionException {
+    public void notify(IEventType event, int variableIdx) throws ContradictionException {
```
[the remark this answers is not in the document]
Comment (cprudhom) [reply solver/src/main/java/org/chocosolver/solver/variables/view/IntView.java:401]: No, that's ok

File: solver/src/main/java/org/chocosolver/solver/variables/view/IView.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cprudhom) [reply solver/src/main/java/org/chocosolver/solver/variables/view/IView.java:48]: That's ok to me like that, we don't need too much abstraction.

**Review by cprudhom — APPROVED** [review]

**Pull request merged**